File: src/main/java/com/indiepost/repository/PostRepository.java
package com.indiepost.repository;

import com.indiepost.dto.post.PostQuery;
import com.indiepost.dto.post.PostSummaryDto;
import com.indiepost.enums.Types;
import com.indiepost.model.Post;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Created by jake on 7/26/16.
 */
public interface PostRepository {

    Post findById(Long id);

    Long count();

    Long count(PostQuery query);

    Types.PostStatus getStatusById(Long postId);

    List<PostSummaryDto> findByIds(List<Long> ids);

    List<PostSummaryDto> findScheduledPosts();

    Page<PostSummaryDto> query(PostQuery postQuery, Pageable pageable);

    Page<PostSummaryDto> findByCategorySlug(String slug, Pageable pageable);

    Page<PostSummaryDto> findByTagName(String tagName, Pageable pageable);

    Page<PostSummaryDto> findByContributorFullName(String fullName, Pageable pageable);

    Page<PostSummaryDto> findByStatus(Types.PostStatus status, Pageable pageable);

    Page<PostSummaryDto> findReadingHistoryByUserId(Long userId, Pageable pageable);

    Page<PostSummaryDto> findBookmarksByUserId(Long userId, Pageable pageable);
}

File: src/main/kotlin/com/indiepost/repository/PostRepository.kt
package com.indiepost.repository

import com.indiepost.dto.post.PostQuery
import com.indiepost.dto.post.PostSummaryDto
import com.indiepost.enums.Types
import com.indiepost.model.Post
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable

/**
 * Created by jake on 7/26/16.
 */
interface PostRepository {

    fun findById(id: Long): Post?

    fun count(): Long

    fun count(postQuery: PostQuery): Long

    fun getStatusById(postId: Long): Types.PostStatus?

    fun findByIds(ids: List<Long>): List<PostSummaryDto>

    fun findScheduledPosts(): List<PostSummaryDto>

    fun query(postQuery: PostQuery, pageable: Pageable): Page<PostSummaryDto>

    fun findByPrimaryTagName(slug: String, pageable: Pageable): Page<PostSummaryDto>

    fun findByTagName(tagName: String, pageable: Pageable): Page<PostSummaryDto>

    fun findByProfileSlug(slug: String, pageable: Pageable): Page<PostSummaryDto>

    fun findPublicPosts(pageable: Pageable, includeFeatured: Boolean = true): Page<PostSummaryDto>

    fun fallbackSearch(text: String, pageable: Pageable): Page<PostSummaryDto>

    fun findRelatedPostsById(id: Long): Page<PostSummaryDto>

    fun findReadingHistoryByUserId(userId: Long, pageable: Pageable): Page<PostSummaryDto>

    fun findBookmarksByUserId(userId: Long, pageable: Pageable): Page<PostSummaryDto>
}